checkQC/handlers/unidentified_index_handler.py

```diff
@@ -0,0 +1,136 @@
+
+from collections import defaultdict
+
+from checkQC.handlers.qc_handler import QCHandler, QCErrorFatal, QCErrorWarning
+from checkQC.parsers.demux_summary_parser import DemuxSummaryParser
+from checkQC.parsers.stats_json_parser import StatsJsonParser
+from checkQC.parsers.samplesheet_parser import SamplesheetParser
+
+
+class UnidentifiedIndexHandler(QCHandler):
+    """
+    TODO
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.lanes_and_indices = {}
+        self.conversion_results = None
+        self.samplesheet = None
+
+    def validate_configuration(self):
+        #TODO Needs custom validation since it does not use same keys as other handlers
+        pass
+
+    def parser(self):
+        return [DemuxSummaryParser, StatsJsonParser, SamplesheetParser]
+
+    def collect(self, signal):
+        if isinstance(signal, tuple):
+            key, value = signal
+            if key == "ConversionResults":
+                self.conversion_results = value
+            if key == "samplesheet":
+                self.samplesheet = value
+        else:
+            self.lanes_and_indices[signal["lane"]] = signal["indices"]
+
+    def check_qc(self):
+
+        samplesheet_dict = self.transform_samplesheet_to_dict(self.samplesheet)
+
+        number_of_reads_per_lane = self.number_of_reads_per_lane()
+        for lane, indices in self.lanes_and_indices.items():
+            number_of_reads_on_lane = number_of_reads_per_lane[lane]
+            for index in indices:
+                tag = index['index']
+                # Unknown index means that the sample was run without an index.
+                if self.is_significantly_represented(index, number_of_reads_on_lane) and not tag == 'unknown':
+                    yield from self.evaluate_index_rules(tag, lane, samplesheet_dict)
+                else:
+                    continue
+
+    def evaluate_index_rules(self, tag, lane, samplesheet_dict):
+        rules = [self.check_swapped_dual_index, self.check_reversed_index,
+                 self.check_reverse_complement_index, self.check_if_index_in_other_lane,
+                 self.check_complement_index]
+        for rule in rules:
+            yield from rule(tag, lane, samplesheet_dict)
+
+    def check_swapped_dual_index(self, tag, lane, samplesheet_dict):
+        if '+' in tag:
+            split_tag = tag.split('+')
+            swapped_tag = '{}+{}'.format(split_tag[1], split_tag[0])
+            hits = self.index_in_samplesheet(samplesheet_dict, swapped_tag)
+            for hit in hits:
+                msg = 'It appears that maybe the dual index tag: {} was swapper. There was a hit for' \
```

Comment:
It would be really useful to include index percentage (that is, how much of the lane does this hit make up for) to the messages.  It makes quite a difference if a hit make up 50% or 1 % of the lane.

Reply:
:+1: